src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py

```diff
@@ -0,0 +1,441 @@
+    48: 245,
+    49: 244,
+    50: 242,
+    51: 241,
+    52: 239,
+    53: 238,
+    54: 236,
+    55: 235,
+    56: 233,
+    57: 232,
+    58: 231,
+    59: 229,
+    60: 228,
+    61: 226,
+    62: 225,
+    63: 223,
+    64: 222,
+    65: 221,
+    66: 220,
+    67: 219,
+    68: 219,
+    69: 218,
+    70: 217,
+    71: 216,
+    72: 216,
+    73: 215,
+    74: 214,
+    75: 213,
+    76: 213,
+    77: 212,
+    78: 211,
+    79: 211,
+    80: 210,
+    81: 209,
+    82: 208,
+    83: 208,
+    84: 207,
+    85: 206,
+    86: 205,
+    87: 205,
+    88: 204,
+    89: 203,
+    90: 202,
+    91: 202,
+    92: 201,
+    93: 200,
+    94: 199,
+    95: 199,
+    96: 198,
+    97: 197,
+    98: 196,
+    99: 196,
+    100: 195,
+    101: 194,
+    102: 193,
+    103: 193,
+    104: 192,
+    105: 191,
+    106: 191,
+    107: 190,
+    108: 189,
+    109: 188,
+    110: 188,
+    111: 187,
+    112: 186,
+    113: 185,
+    114: 185,
+    115: 184,
+    116: 183,
+    117: 182,
+    118: 182,
+    119: 181,
+    120: 180,
+    121: 179,
+    122: 179,
+    123: 178,
+    124: 177,
+    125: 176,
+    126: 176,
+    127: 175,
+    128: 174,
+}
+
+MAX_DISCOUNT = 174
+MULTIPLIER = 1000
+
+
+def bytes_to_G1(data: Bytes) -> Point2D:
+    """
+    Decode 128 bytes to a G1 point. Does not perform sub-group check.
+
+    Parameters
+    ----------
+    data :
+        The bytes data to decode.
+
+    Returns
+    -------
+    point : Point2D
+        The G1 point.
+
+    Raises
+    ------
+    InvalidParameter
+        Either a field element is invalid or the point is not on the curve.
+    """
+    assert len(data) == 128
+    x = int.from_bytes(data[:64], "big")
+    y = int.from_bytes(data[64:], "big")
+
+    if x >= P:
+        raise InvalidParameter("Invalid field element")
+    if y >= P:
+        raise InvalidParameter("Invalid field element")
+
+    if x == 0 and y == 0:
+        return None
+
+    point = (FQ(x), FQ(y))
+
+    # Check if the point is on the curve
+    if not is_on_curve(point, b):
+        raise InvalidParameter("Point is not on curve")
+
+    return point
+
+
+def G1_to_bytes(point: Point2D) -> Bytes:
+    """
+    Encode a G1 point to 128 bytes.
+
+    Parameters
+    ----------
+    point :
+        The G1 point to encode.
+
+    Returns
+    -------
+    data : Bytes
+        The encoded data.
+    """
+    if point is None:
+        return b"\x00" * 128
+
+    x, y = point
+
+    x_bytes = int(x).to_bytes(64, "big")
+    y_bytes = int(y).to_bytes(64, "big")
+
+    return x_bytes + y_bytes
+
+
+def decode_G1_scalar_pair(data: Bytes) -> Tuple[Point2D, int]:
+    """
+    Decode 160 bytes to a G1 point and a scalar.
+
+    Parameters
+    ----------
+    data :
+        The bytes data to decode.
+
+    Returns
+    -------
+    point : Tuple[Point2D, int]
+        The G1 point and the scalar.
+
+    Raises
+    ------
+    InvalidParameter
+        If the sub-group check failed.
+    """
+    assert len(data) == 160
```

Comment:
Probably shouldn't be an assertion:

```suggestion
    if len(data) != 160:
        raise SomeException()
```